tests/bluetooth/audio/cap_commander/uut/bap_broadcast_assistant.c

```diff
@@ -6,24 +6,160 @@
 
 #include "zephyr/bluetooth/audio/bap.h"
 
-static struct bt_bap_broadcast_assistant_cb *broadcast_assistant_cbs;
+static sys_slist_t broadcast_assistant_cbs = SYS_SLIST_STATIC_INIT(&broadcast_assistant_cbs);
+
+struct bap_broadcast_assistant_recv_state_info {
+	uint8_t src_id;
+	/** Cached PAST available */
+	bool past_avail;
+	uint8_t adv_sid;
+	uint32_t broadcast_id;
+	bt_addr_le_t addr;
+};
+
+struct bap_broadcast_assistant_instance {
+	struct bt_conn *conn;
+	struct bap_broadcast_assistant_recv_state_info
+		recv_states[CONFIG_BT_BAP_BROADCAST_ASSISTANT_RECV_STATE_COUNT];
+	bool inst_active;
+};
+
+static struct bap_broadcast_assistant_instance broadcast_assistants[CONFIG_BT_MAX_CONN];
+static uint8_t max_src_id;
+
+static struct bap_broadcast_assistant_instance *inst_by_conn(struct bt_conn *conn)
+{
+	struct bap_broadcast_assistant_instance *inst;
+
+	zassert_not_null(conn, "conn is NULL");
+
+	inst = &broadcast_assistants[bt_conn_index(conn)];
+
+	return inst;
+}
 
 int bt_bap_broadcast_assistant_register_cb(struct bt_bap_broadcast_assistant_cb *cb)
 {
-	broadcast_assistant_cbs = cb;
+	struct bt_bap_broadcast_assistant_cb *tmp;
+
+	if (cb == NULL) {
+		return -EINVAL;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&broadcast_assistant_cbs, tmp, _node) {
+		if (tmp == cb) {
+			return -EALREADY;
+		}
+	}
+
+	sys_slist_append(&broadcast_assistant_cbs, &cb->_node);
+
+	return 0;
+}
+
+int bt_bap_broadcast_assistant_unregister_cb(struct bt_bap_broadcast_assistant_cb *cb)
+{
+	if (cb == NULL) {
+		return -EINVAL;
+	}
+
+	if (!sys_slist_find_and_remove(&broadcast_assistant_cbs, &cb->_node)) {
+		return -EALREADY;
+	}
 
 	return 0;
 }
 
 int bt_bap_broadcast_assistant_add_src(struct bt_conn *conn,
 				       const struct bt_bap_broadcast_assistant_add_src_param *param)
 {
+	struct bap_broadcast_assistant_instance *inst;
+	struct bt_bap_scan_delegator_recv_state state;
+	struct bt_bap_broadcast_assistant_cb *listener, *next;
+
 	/* Note that proper parameter checking is done in the caller */
 	zassert_not_null(conn, "conn is NULL");
 	zassert_not_null(param, "param is NULL");
 
-	if (broadcast_assistant_cbs->add_src != NULL) {
-		broadcast_assistant_cbs->add_src(conn, 0);
+	inst = inst_by_conn(conn);
+	zassert_not_null(inst, "inst is NULL");
+
+	max_src_id++;
+	inst->recv_states[0].src_id = max_src_id;
+	inst->recv_states[0].past_avail = false;
+	inst->recv_states[0].adv_sid = param->adv_sid;
+	inst->recv_states[0].broadcast_id = param->broadcast_id;
+	inst->inst_active = true;
+	state.pa_sync_state = param->pa_sync;
+	state.src_id = max_src_id;
+	state.num_subgroups = param->num_subgroups;
+	for (size_t i = 0; i < param->num_subgroups; i++) {
+		state.subgroups[i].bis_sync = param->subgroups[i].bis_sync;
+	}
+
+	bt_addr_le_copy(&inst->recv_states[0].addr, &param->addr);
```

Comment:
We should either access the correct `recv_states` instead of always `0`, or not have it as an array.